### PR TITLE
Extract auth module

### PR DESCRIPTION
**Pull request by brancz**
This pull request extracts the authentication parts with Kubernetes into a re-usable module to be importable by other projects.

@s-urbaniak @squat @mxinden

### CODE REVIEW

File: auth/auth.go [hunk 1]
```diff
@@ -36,6 +36,14 @@ import (
 	authorizationclient "k8s.io/client-go/kubernetes/typed/authorization/v1beta1"
 )
 
+type AuthInterface interface {
```
Comment (s-urbaniak) [inline auth/auth.go]: nit `AuthInterface interface` stutters, can we call this `Authorizer`?

Comment (brancz) [reply auth/auth.go]: It encapsulates both authentication and authorization, so `Authorizer` would not give the right notion, I think.

Comment (squat) [reply auth/auth.go]: I agree with Frederic. Like anything there can be exceptions to these naming guidelines. One thing is I always find it funny when types have the type in the name. We could call this simply `Auther`.

File: auth/auth.go [hunk 2]
```diff
@@ -76,6 +84,58 @@ type AuthConfig struct {
 	Authorization  *AuthzConfig
 }
 
+func (c *AuthConfig) DeepCopy() *AuthConfig {
+	res := &AuthConfig{
+		Authentication: &AuthnConfig{
+			X509: &X509Config{
+				ClientCAFile: c.Authentication.X509.ClientCAFile,
+			},
+			Header: &AuthnHeaderConfig{
```
[the remark this answers is not in the document]
Comment (brancz) [reply auth/auth.go]: You're absolutely right. I think on line 89 we should just not create any deep structures and just build them as we already do later in the code. Nice catch!

[the remark this answers is not in the document]
Comment (brancz) [reply auth/auth.go]: Fixed.

File: auth/auth.go (hunk 1 above)
Comment (s-urbaniak) [reply auth/auth.go]: If it encapsulates two separate concerns, it smells like the interface is too big.

Just asking: Do we need to expose this interface at all? Why not simply returning `(*kubeRBACProxyAuth, error)`? The interface declaration could be anonymous, ensuring compile time safety, the caller could then type cast to `Authorizer` and/or `Authenticator` at will.

All of this is type bike shedding though, ignore it at will ;-)

Comment (brancz) [reply auth/auth.go]: I like `Auther`, let's go with that.

Comment (brancz) [reply auth/auth.go]: @s-urbaniak agreed. I removed it.

**Pull request merged**